src/algorithms/local_search/swap_star_utils.h

```diff
@@ -0,0 +1,353 @@
+
+struct InsertionRange {
+  std::vector<Index> range;
+  Index first_rank;
+  Index last_rank;
+};
+
+// Compute insertion range in source route when removing job at s_rank
+// and adding job at job_rank in source route at insertion_rank.
+inline InsertionRange get_insert_range(const std::vector<Index>& s_route,
+                                       Index s_rank,
+                                       Index job_rank,
+                                       Index insertion_rank) {
+  InsertionRange insert;
+  if (s_rank == insertion_rank) {
+    insert.range.push_back(job_rank);
+    insert.first_rank = s_rank;
+    insert.last_rank = s_rank + 1;
+  } else {
+    if (s_rank < insertion_rank) {
+      std::copy(s_route.begin() + s_rank + 1,
+                s_route.begin() + insertion_rank,
+                std::back_inserter(insert.range));
+      insert.range.push_back(job_rank);
+      insert.first_rank = s_rank;
+      insert.last_rank = insertion_rank;
+    } else {
+      insert.range.push_back(job_rank);
+      std::copy(s_route.begin() + insertion_rank,
+                s_route.begin() + s_rank,
+                std::back_inserter(insert.range));
+      insert.first_rank = insertion_rank;
+      insert.last_rank = s_rank + 1;
+    }
+  }
+
+  return insert;
+}
+
+template <class Route>
+SwapChoice compute_best_swap_star_choice(const Input& input,
+                                         const utils::SolutionState& sol_state,
+                                         const Route& source,
+                                         const Route& target,
+                                         Gain best_known_gain) {
+  // Preprocessing phase.
+  std::unordered_map<Index, ThreeInsertions> top_insertions_in_target;
+  for (unsigned s_rank = 0; s_rank < source.route.size(); ++s_rank) {
+    const auto source_job_rank = source.route[s_rank];
+
+    if (input.jobs[source_job_rank].type == JOB_TYPE::SINGLE and
+        input.vehicle_ok_with_job(target.vehicle_rank, source_job_rank)) {
+      top_insertions_in_target[s_rank] =
+        find_top_3_insertions(input, source_job_rank, target);
+    }
+  }
+
+  std::unordered_map<Index, ThreeInsertions> top_insertions_in_source;
+  for (unsigned t_rank = 0; t_rank < target.route.size(); ++t_rank) {
+    const auto target_job_rank = target.route[t_rank];
+
+    if (input.jobs[target_job_rank].type == JOB_TYPE::SINGLE and
+        input.vehicle_ok_with_job(source.vehicle_rank, target_job_rank)) {
+      top_insertions_in_source[t_rank] =
+        find_top_3_insertions(input, target_job_rank, source);
+    }
+  }
+
+  // Search phase.
+  auto best_choice = empty_choice;
+  Gain best_gain = best_known_gain;
+
+  for (const auto& s_element : top_insertions_in_target) {
+    const auto s_rank = s_element.first;
+    const auto& target_insertions = s_element.second;
+
+    const auto source_remove_gain =
+      sol_state.node_gains[source.vehicle_rank][s_rank];
```

Comment:
There is a flaw here because what the algorithm requires is a cost delta. Namely if job `j` has previous and next steps `p` and `n`, then `Delta(j, p, n) = cost(p, j) + cost(j, n) - cost(p, n)`.

This is exactly what `sol_state.node_gains` holds in most situations. The problem is if the route has length 1 and vehicle has start and end that differ. Then for the delta computation `cost(p, n)` should be the cost of getting from start to end in an empty route. This is not accounted for in `sol_state.node_gains` because we consider that an empty route does not generate a start -> end cost.

Currently this leads to wrong gain estimates with a SWAP* move involving such a route, which hits the assert on gain verification we have down the line after applying the moves.